fix: Server-side diff shows incorrect diffs for list related changes

pjiang-dev opened the pull request:
fixes https://github.com/argoproj/argo-cd/issues/21817

The current Difference function does not remove every field and its children from the set. This causes unwanted the behavior of nested fields not being removed resulting in an incorrect server side diff. Using the RecursiveDifference function instead will do this:
"this removes every field **and its children** from s that is contained in s2."

Added additional unit test to cover the nested fields scenario.
Before fix:
<img width="1400" alt="Screenshot 2025-02-12 at 10 55 32 AM" src="https://github.com/user-attachments/assets/0b35d801-e151-4b4c-8b60-70555b1c3b8a" />


After fix:
<img width="1079" alt="Screenshot 2025-02-12 at 10 55 10 AM" src="https://github.com/user-attachments/assets/8b964457-aaa0-43c1-9bb9-08d961cb252d" />

andrii-korotkov-verkada commented:
Ah, looks like cloud doesn't support blobs https://docs.sonarsource.com/sonarqube-server/latest/project-administration/analysis-scope/#wildcard-examples. I can't find exact explanation what that means, but would try replacing `pkg/diff/testdata/**` with `pkg/diff/testdata/**/*` as well.

pjiang-dev commented:
> Ah, looks like cloud doesn't support blobs https://docs.sonarsource.com/sonarqube-server/latest/project-administration/analysis-scope/#wildcard-examples. I can't find exact explanation what that means, but would try replacing `pkg/diff/testdata/**` with `pkg/diff/testdata/**/*` as well.

Strange, tried a couple different wildcards and they don't exclude the testdata files

andrii-korotkov-verkada commented:
Did you try adding spaces around = ? Tho I swear it worked for me before just like it was in master.

andrii-korotkov-verkada commented:
Hm, now I can only think of two things:
* Sonar config usage is broken.
* Files need to be excluded individually.

zachaller commented:
I would imagine part of this issue is with Lists only due to not having merge keys, does RecursiveDifference only do that whole replacement on lists or does it do it for every object type? If it does it for every object type I am not sure that is the correct action but I will differ to @leoluz

pjiang-dev commented:
> I would imagine part of this issue is with Lists only due to not having merge keys, does RecursiveDifference only do that whole replacement on lists or does it do it for every object type? If it does it for every object type I am not sure that is the correct action but I will differ to @leoluz

This RecursiveDIfference only works on sets. This is coming from the kubernetes structured-diff library where the original DIfference comes from.

pjiang-dev commented:
There is a bug using this approach when manifest files have a field added by mutation webhooks:
![Screenshot 2025-02-12 at 1 58 22 PM](https://github.com/user-attachments/assets/8b585797-cbf3-4a29-b2d7-d375fd955dfa)

Basically RecursiveDIfference will remove fields even if it is not an exact match. It works by removing subtrees only when there's an exact match at a non-leaf level.

Example:
Added by mutation webhooks set:
```
.metadata.creationTimestamp
.metadata.labels.peter-test
```

Managed Fields Set:
`.metadata.labels.app.kubernetes.io/instance`

In this case `.metadata.labels.peter-test` will be removed because it matches the field `.metadata.labels` inside the managed fields set, which is incorrect.

One option is to create a custom Difference function to remove fields at a more granular level.

andrii-korotkov-verkada commented:
https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/fieldpath/set.go#L110
Maybe we can ask the community to implement the right variation of the function.

andrii-korotkov-verkada commented:
@leoluz, do you think we should implement a custom function or ask structured merge diff devs to do it? I feel like the 2nd is better, though would take some time.

andrii-korotkov-verkada commented:
Shall we only remove mfs for leaf nodes or something like that?

andrii-korotkov-verkada commented on pkg/diff/diff.go:
```diff
@@ -236,15 +236,15 @@ func removeWebhookMutation(predictedLive, live *unstructured.Unstructured, gvkPa
 		if comparison.Added != nil && !comparison.Added.Empty() {
 			// exclude the added fields owned by this manager from the comparison
-			comparison.Added = comparison.Added.Difference(mfs)
+			comparison.Added = comparison.Added.RecursiveDifference(mfs)
```
Or, I wonder if we can do something like
```
predictedLive.ExtractItems(plManagedFields, typed.WithAppendKeyFields())
```

andrii-korotkov-verkada replied:
Or maybe we just remove every prefix of every managed field path.

pjiang-dev replied:
I tried this as well, and it results to messed up outputs.

andrii-korotkov-verkada replied:
Ah, nvm, I think we should use RecursiveDifference, but only with leaf mfs. What do you think?

andrii-korotkov-verkada replied:
One more thought - also remove managed fields from the current live.

pjiang-dev replied:
I think we can't use RecursiveDifference because it could incorrectly remove fields from mutation webhooks. 
However, yes, I think what we could do is keep Difference, but do a second pass to remove any leaf nodes that are still in managed fields that weren't caught by Difference.

pjiang-dev replied:
I will try out this and also removing manged fields from current live and see if tit works

andrii-korotkov-verkada commented:
> In this case .metadata.labels.peter-test will be removed because it matches the field .metadata.labels inside the managed fields set, which is incorrect.

Why would `.metadata.labels` be a part of managed fields set if not all of it is managed?

andrii-korotkov-verkada commented:
One more thought - do a recursive difference, but only if the field is managed in both live and predicted live.

andrii-korotkov-verkada commented:
```
// RecursiveDifference returns a Set containing elements which:
// * appear in s
// * do not appear in s2
//
// Compared to a regular difference,
// this removes every field **and its children** from s that is contained in s2.
//
// For example, with s containing `a.b.c` and s2 containing `a.b`,
// a RecursiveDifference will result in `a`, as the entire node `a.b` gets removed.
```
In the example
```
Added by mutation webhooks set:

.metadata.creationTimestamp
.metadata.labels.peter-test

Managed Fields Set:
.metadata.labels.app.kubernetes.io/instance

In this case .metadata.labels.peter-test will be removed because it matches the field .metadata.labels inside the managed fields set, which is incorrect.
```
I don't think `.metadata.labels.peter-test` would be removed, unless `.metadata.labels` is also in the managed field set (perhaps because predicted live only has `.metadata.labels.app.kubernetes.io/instance`.

andrii-korotkov-verkada commented:
So IIUC in the example above predicted live should contain:
```
.metadata.creationTimestamp
.metadata.labels.peter-test
.metadata.labels.app.kubernetes.io/instance
```
But I don't see why `.metadata.labels` would be in the managed fields set.

andrii-korotkov-verkada commented:
Do I understand correctly, that the problem is with partially managed fields (with children) being considered as fully managed?

andrii-korotkov-verkada commented:
Can you also try just this modification, still with using Difference?

Now:
```
	if comparison.Added != nil && !comparison.Added.Empty() {
		typedPredictedLive = typedPredictedLive.RemoveItems(comparison.Added)
	}
```
Desired?
```
	if comparison.Added != nil && !comparison.Added.Empty() {
                // Might extract some key fields still used
                liveAddValues := typedLive.ExtractItems(comparison.Added, typed.WithAppendKeyFields()) 
		typedPredictedLive = typedPredictedLive.RemoveItems(comparison.Added)
                typedPredictedLive = typedPredictedLive.Merge(liveAddValues)
	}
```

pjiang-dev commented:
Here is what's happening exactly:

Before Difference() on Line 239
mfs:
```
.metadata.labels.app.kubernetes.io/instance
.spec.selector
.spec.template.metadata.labels.app.kubernetes.io/name
.spec.template.spec.containers[name="idle"]
.spec.template.spec.containers[name="idle"].image
.spec.template.spec.containers[name="idle"].name
.spec.template.spec.containers[name="idle"].env[name="FOO"]
.spec.template.spec.containers[name="idle"].env[name="FOO"].name
.spec.template.spec.containers[name="idle"].env[name="FOO"].value
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"]
.spec.template.spec.containers[name="idle"].ports[containerPort=9876,protocol="TCP"]
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].containerPort
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].name
.spec.template.spec.containers[name="idle"].ports[containerPort=9876,protocol="TCP"].containerPort
.spec.template.spec.containers[name="idle"].ports[containerPort=9876,protocol="TCP"].name
```

Comparison:
```
- Modified Fields:
.metadata.generation
.metadata.managedFields
- Added Fields:
.metadata.creationTimestamp
.metadata.labels.peter-test
.spec.template.spec.containers[name="idle"].env
.spec.template.spec.containers[name="idle"].env[name="FOO"]
.spec.template.spec.containers[name="idle"].env[name="FOO"].name
.spec.template.spec.containers[name="idle"].env[name="FOO"].value
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"]
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].containerPort
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].name
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].protocol
```

After Difference() on line 239
Comparison:
```
- Modified Fields:
.metadata.generation
.metadata.managedFields
- Added Fields:
.metadata.creationTimestamp
.metadata.labels.peter-test
.spec.template.spec.containers[name="idle"].env
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].protocol
```

The bug is that these two fields remain in comparison.Added even after the Difference function even though they are managed fields. If they are in comparison.Added they will be extracted from predictedLive and cause the diff to not show these fields. 
```
.spec.template.spec.containers[name="idle"].env
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].protocol
```

The cause is the Difference function will does a shallow difference and does not recursively check each child.

pjiang-dev commented:
> Can you also try just this modification, still with using Difference?
> 
> Now:
> 
> ```
> 	if comparison.Added != nil && !comparison.Added.Empty() {
> 		typedPredictedLive = typedPredictedLive.RemoveItems(comparison.Added)
> 	}
> ```
> 
> Desired?
> 
> ```
> 	if comparison.Added != nil && !comparison.Added.Empty() {
>                 // Might extract some key fields still used
>                 liveAddValues := typedLive.ExtractItems(comparison.Added, typed.WithAppendKeyFields()) 
> 		typedPredictedLive = typedPredictedLive.RemoveItems(comparison.Added)
>                 typedPredictedLive, _ = typedPredictedLive.Merge(liveAddValues)
> 	}
> ```

This one also did not make a difference. I think because we are still extracting those fields.

andrii-korotkov-verkada commented:
Great analysis!

> .spec.template.spec.containers[name="idle"].env
Looks like fields that are parents of other fields are not managed themselves.
We run into a tricky situation when they were added, but only some of their children are managed.
In such case, after removing managed fields, these kind of fields would still remain, so we shouldn't touch them.
In other words, we should remove a non-managed field if and only if all its children are managed.

> .spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].protocol
This is due to a key field not specified explicitly, but it can be derived from the path.
It's actually an easier case, since its parent is managed.
We should remove a non-managed field if its parent subpath is managed (e.g. `.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"]`).

andrii-korotkov-verkada commented:
The solution is:
We compute the leaves of added fields before the mfs subtraction loop. After the loop we go through the remaining added fields and add some fields to a new excluded field set:
* If it's a leaf, we add all of its subpaths to this field set.
* If any of its subpaths is managed, we add the path itself to this set.
Then we subtract this field set from the added fields.

pjiang-dev commented:
@andrii-korotkov-verkada I think this is a good solution. Basically it covers the two edge cases 
Case 1: Parent fields with managed children
.spec.template.spec.containers[name="idle"].env
(children like .env[name="FOO"] are managed)

Case 2: Derived fields with managed parents
.spec.template.spec.containers[name="idle"].ports[containerPort=3000,protocol="TCP"].protocol
(parent path is managed)

I can start implementing this and @leoluz can confirm if it makes sense
